src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,162 @@
+    public List<Task> load() throws DukeException {
+        File file = readFile();
+        assert file != null;
+        return parseFile(file);
+    }
+
+    /**
+     * Reads the file from the file at location filePath.
+     *
+     * @return File located at location filePath.
+     * @throws DukeException Exception when there are errors in file reading.
+     */
+    public File readFile() throws DukeException {
+        try {
+            File data = new File(filePath);
+            data.getParentFile().mkdirs();
+            data.createNewFile();
+            return data;
+        } catch (IOException e) {
+            throw new DukeIoException();
+        }
+    }
+
+
+    /**
+     * ReadLine takes in a String and returns a Task.
+     *
+     * @param line String containing information about a Task.
+     * @return Task which corresponds to the String line.
+     * @throws DukeException Exception when there are errors in reading the line.
+     */
+    public Task readLine(String line) throws DukeException {
+        Pattern todoPattern = Pattern.compile("T \\| ([01]) \\| (.+)");
+        Pattern deadlinePattern = Pattern.compile("D \\| ([01]) \\| (.+?) \\| (.+)");
+        Pattern eventPattern = Pattern.compile("E \\| ([01]) \\| (.+?) \\| (.+)");
+        Matcher todoMatcher = todoPattern.matcher(line);
+        Matcher deadlineMatcher = deadlinePattern.matcher(line);
+        Matcher eventMatcher = eventPattern.matcher(line);
+
+        if (todoMatcher.find()) {
+            return new TodoTask(todoMatcher.group(2), Integer.parseInt(todoMatcher.group(1)));
+        } else if (deadlineMatcher.find()) {
+            return new DeadlineTask(
+                    deadlineMatcher.group(2),
+                    Integer.parseInt(deadlineMatcher.group(1)),
+                    LocalDateTime.parse(deadlineMatcher.group(3)));
+        } else if (eventMatcher.find()) {
+            return new EventTask(
+                    eventMatcher.group(2),
+                    Integer.parseInt(eventMatcher.group(1)),
+                    LocalDateTime.parse(eventMatcher.group(3)));
+        } else {
+            // Formats do not match, input given is invalid
+            throw new InvalidInputException();
+        }
+    }
+
+    /**
+     * Parses the contents of the file given into a list of tasks.
+     *
+     * @param file File that needs to parsed.
+     * @return ArrayList containing Task Objects.
+     * @throws DukeException Exceptions in parsing the file.
+     */
+    public List<Task> parseFile(File file) throws DukeException {
+        List<Task> results = new ArrayList<>();
+        try {
+            BufferedReader reader = new BufferedReader(new FileReader(file));
+            String line;
+            // T | 1 | read book
+
+            while ((line = reader.readLine()) != null) {
+                Task task = readLine(line);
+                if (task != null) {
+                    results.add(task);
+                }
+            }
```

Comment:
Any reason why you chose to combine the variable assignment and checking `!= null` within the `while` condition, instead of keeping them separate? In my view, the second option might improve readability.

Reply:
I have amended the code to follow the second option, the code looks easier to read now, thanks.